core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -1334,22 +1334,28 @@ public RelBuilder filter(Iterable<CorrelationId> variablesSet,
    *
    * <p>The predicates are combined using AND,
    * and optimized in a similar way to the {@link #and} method.
-   * If the result is TRUE no filter is created. */
+   * If simplification is on and the result is TRUE, no filter is created. */
   public RelBuilder filter(Iterable<CorrelationId> variablesSet,
       Iterable<? extends RexNode> predicates) {
-    final RexNode simplifiedPredicates =
-        simplifier.simplifyFilterPredicates(predicates);
-    if (simplifiedPredicates == null) {
-      return empty();
+    final RexNode conjunctionPredicates;
+    if (config.simplify()) {
+      conjunctionPredicates = simplifier.simplifyFilterPredicates(predicates);
+      if (conjunctionPredicates == null) {
+        return empty();
+      }
+      if (conjunctionPredicates.isAlwaysTrue()) {
+        return this;
+      }
```

Comment:
I think this nested ifs should not be here but outside the main if (as it was before) cause otherwise we might be changing unintentionally the old behavior.

Reply:
Sorry, not get your point😂, can you desc more?

Reply:
Move lines 1343 to 1348 after 1352.

Reply:
Nice point, this will still do simplification, but I think this degree of optimization is acceptable.